packages/wallet/src/TxBuilder/buildTx.ts

```diff
@@ -0,0 +1,142 @@
+  auxiliaryData
+}: CreateSignedTxProps): Promise<SignedTx> => {
+  // REVIEW: we have to update the ObservableWallet interface `finalizeTx` method
+  // to include options?: SignTransactionOptions. Otherwise there's no way to pass the `props` to finalizeTx
+  props; // TODO mh: remove me
+  const tx = await wallet.finalizeTx(initializeTxResult, auxiliaryData);
+  return {
+    submit: () => wallet.submitTx(tx),
+    tx
+  };
+};
+
+class TxBuilderSaw implements TxBuilder {
+  partialTxBody: Partial<Cardano.NewTxBodyAlonzo> = {};
+  auxiliaryData?: Cardano.AuxiliaryData;
+  extraWitness?: Partial<Cardano.Witness>;
+
+  #observableWallet: TxBuilderObservableWallet;
+  #util: OutputValidator;
+
+  constructor(observableWallet: TxBuilderObservableWallet, util: OutputValidator) {
+    this.#observableWallet = observableWallet;
+    this.#util = util;
+  }
+
+  addOutput(txOut: Cardano.TxOut): TxBuilder {
+    this.partialTxBody.outputs = [...(this.partialTxBody.outputs || []), txOut];
+    return this;
+  }
+
+  removeOutput(txOut: Cardano.TxOut): TxBuilder {
+    this.partialTxBody.outputs = this.partialTxBody.outputs?.filter((output) => !deepEquals(output, txOut));
+    return this;
+  }
+
+  buildOutput(txOut?: DeepPartial<Cardano.TxOut>): OutputBuilder {
+    throw new NotImplementedError(`TODO + ${txOut}`);
+  }
+
+  delegate(poolId: Cardano.PoolId): TxBuilder {
+    throw new NotImplementedError(`TODO + ${poolId}`);
+  }
+
+  setMetadata(metadata: Cardano.TxMetadata): TxBuilder {
+    throw new NotImplementedError(`TODO + ${metadata}`);
+  }
+
+  async build(): Promise<MaybeValidTx> {
+    try {
+      await this.#validateOutputs();
+
+      // TODO mh: initializeTx should throw TxBodyValidationErrors
+      // validate outputs (e.g. use wallet utils -- token bundle size, min utxo value that can be sent on output)
+      const initializeTxResult = await this.#observableWallet.initializeTx({
+        outputs: new Set(this.partialTxBody.outputs || [])
+      });
+      return {
+        body: initializeTxResult.body,
+        isValid: true,
+        sign: (props?: SignTransactionOptions) =>
+          createSignedTx({
+            auxiliaryData: this.auxiliaryData,
+            initializeTxResult,
+            props,
+            wallet: this.#observableWallet
+          })
+      };
+    } catch (error) {
+      const errors = Array.isArray(error) ? (error as TxBodyValidationError[]) : [error as TxBodyValidationError];
+      return {
+        errors,
+        isValid: false
+      };
+    }
+  }
+
+  // REVIEW: we could make this public and allow users to run validations at any time.
+  async #validateOutputs(): Promise<OutputValidationError[]> {
+    const outputValidations =
+      this.partialTxBody.outputs && (await this.#util.validateOutputs(this.partialTxBody.outputs));
+
+    const errors = [...(outputValidations?.values() || [])]
+      .map((validation) => {
+        // REVIEW: Should we add the whole validation data, for more context
```

Comment:
Yep, and the invalid output obj as well. Should probably be part of OutputValidationError. Might need to split it into 2 error types then?